### DISCUSSION TITLE
Cache Routes to allow for Tabs to keep alive while offscreen

### DISCUSSION THREAD

**Discussion by ziongh** [TanStack/router#1447] [Ideas]
It would be awesome to have a functionality to allow multiple routes to be opened simultaneously so that the user can navigate between the components without unmounting and remounting them. This would allow for an Application to have tabs functionality similar to chrome:

![image](https://github.com/TanStack/router/assets/5855426/0b5a310c-8aaa-4704-8b54-587caffda253)
-- image from: https://github.com/Southliu/react-admin

I don't known the best way to achieve that, There could be a hidden component that when the page goes offscreen it holds a reference to the page, and when a navigation happens, it first looks for this offscreen "cache" to known if there is already a live component, otherwise load the component.

Maybe there is a better option, but this would be great, because the user wouldn't need to open multiple tabs, and this would decrease memory comsumption.

**Comment by Atw-Lee** [comment]
👀

**Comment by recallwei** [comment]
I also need this feature.

**Comment by ludersGabriel** [comment]
This would be amazing

**Comment by ArcherCube** [comment]
maybe you can see #823 .
I have tried the cache plan (it essentially deep clone the router context value and create a new RouterContext provider for the outlet inside), seems work, but haven't tested.

**Reply by BrendanC23** [reply]
Can you share your code?

**Reply by ArcherCube** [reply]
sure.
```tsx
import { getRouterContext, Outlet } from '@tanstack/react-router';
import cloneDeep from 'lodash-es/cloneDeep';
import { useContext, useEffect, useMemo } from 'react';

export const Content = () => {
  const { activeKey, openedPages, setOpenedPages } = useContext(LayoutContext);

  const RouterContext = useMemo(() => getRouterContext(), [activeKey]);
  const routerContextValue = useContext(RouterContext);

  useEffect(() => {
    if (!openedPages[activeKey]) {
      const currentContextValue = cloneDeep(routerContextValue);
      setOpenedPages((current) => {
        return {
          ...current,
          [activeKey]: {
            // cache the outlet wrapped by context 
            element: (
              <RouterContext.Provider value={currentContextValue}>
                <Outlet />
              </RouterContext.Provider>
            ),
          },
        };
      });
    }
  }, [openedPages, setOpenedPages, activeKey, routerContextValue, RouterContext]);

  return (
    <div>
      {Object.entries(openedPages).map(([pageKey, page]) => {
        const isCurrent = activeKey === pageKey;
        return (
          <div
            className={cn('relative', isCurrent ? 'block' : 'hidden')}
            key={pageKey}
          >
            {page.element}
          </div>
        );
      })}
    </div>
  )
}
```

**Reply by lenran659** [reply]
Hi! I tried your code. There still seem to be some problems. When I switch from Route A to Route B, the page still shows what's on Route A.

**Reply by ArcherCube** [reply]
> Hi! I tried your code. There still seem to be some problems. When I switch from Route A to Route B, the page still shows what's on Route A.

maybe you could share your code.
In my case, I render all of the cached route components and just simply set the active one to display: block, while all others are set to display: hidden.

**Comment by hemengke1997** [comment]
@ArcherCube Thank you for sharing! I've created a KeepAlive library that makes it incredibly easy to implement off-screen rendering functionality similar to Vue.js, and it's very straightforward to use!
See: [tanstack-router-keepalive](https://github.com/hemengke1997/tanstack-router-keepalive)

**Reply by qaez0** [reply]
Hello ? i tried your library it works. But im having problem with useSearch. i needed my search params, it does not change value making my useQuery not fire as it used to incase of page change or other api stuff. Url changed but the value of useSearch did not Is there a way to make this work? and also the documentation is not updated.

**Reply by ludersGabriel** [reply]
the lib is pretty cool @hemengke1997 although it does have the issue with params

**Comment by ludersGabriel** [comment]
@ziongh did u find a way to do this properly?
